File: algorithms/[160]相交链表/solution.py
# topics = ["链表", "哈希表"]

from __future__ import annotations
from typing import Union, Set


class ListNode:
    def __init__(self, x: int = 0):
        self.val = x
        self.next: Union[ListNode, None] = None


class Solution:
    def getIntersectionNode(
        self, headA: Union[ListNode, None], headB: Union[ListNode, None]
    ) -> ListNode:
        """
        Hash Table
        time O(m + n), space O(m), m 和 n 分别为 headA 和 headB 的长度
        """
        setA: Set[ListNode] = set()

        while headA:
            setA.add(headA)
            headA = headA.next

        while headB:
            if headB in setA:
                return headB
            headB = headB.next

        return ListNode()

File: algorithms/[160]相交链表/solution.ts
// topics = ["链表", "双指针"]

export class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

/**
 * Two Pointers
 * time O(m + n), space O(1), m 和 n 分别为 headA 和 headB 的长度
 */
function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
  if (!headA || !headB) {
    return null;
  }
  // 当 pA 到达链表的尾部时，将它重定位到链表 B 的头结点
  // 当 pB 到达链表的尾部时，将它重定位到链表 A 的头结点
  // 参考 https://leetcode-cn.com/problems/intersection-of-two-linked-lists/solution/tu-jie-xiang-jiao-lian-biao-by-user7208t/
  let pA: ListNode | null = headA;
  let pB: ListNode | null = headB;
  while (pA !== pB) {
    pA = pA === null ? headB : pA.next;
    pB = pB === null ? headA : pB.next;
  }

  // 不相交时为 null
  return pA;
}
